README.rst
.. image:: https://github.com/AdCombo/flask-combo-jsonapi/workflows/Python%20tests%20and%20coverage/badge.svg
   :alt: flask-combo-jsonapi actions
   :target: https://github.com/AdCombo/flask-combo-jsonapi/actions
.. image:: https://coveralls.io/repos/github/AdCombo/flask-combo-jsonapi/badge.svg
   :alt: flask-combo-jsonapi coverage
   :target: https://coveralls.io/github/AdCombo/flask-combo-jsonapi
.. image:: https://img.shields.io/pypi/v/flask-combo-jsonapi.svg
   :alt: PyPI
   :target: http://pypi.org/p/flask-combo-jsonapi


Flask-COMBO-JSONAPI
###################

Flask-COMBO-JSONAPI is a flask extension for building REST APIs.
It combines the power of `Flask-Restless`_
and the flexibility of `Flask-RESTful`_
around a strong specification `JSONAPI 1.0 <http://jsonapi.org/>`_.
This framework is designed to quickly build REST APIs and fit the complexity
of real life projects with legacy data and multiple data storages.

The main goal is to make it flexible using `plugin system <https://combojsonapi.readthedocs.io/>`_


Install
=======

    pip install Flask-COMBO-JSONAPI


A minimal API
=============

.. code-block:: python

    from flask import Flask
    from flask_combo_jsonapi import Api, ResourceDetail, ResourceList
    from flask_sqlalchemy import SQLAlchemy
    from marshmallow import pre_load
    from marshmallow_jsonapi.flask import Schema
    from marshmallow_jsonapi import fields

    # Create the Flask application and the Flask-SQLAlchemy object.
    app = Flask(__name__)
    app.config['DEBUG'] = True
    app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:////tmp/api_minimal.db'
    db = SQLAlchemy(app)


    # Create model
    class Person(db.Model):
        id = db.Column(db.Integer, primary_key=True)
        name = db.Column(db.String)


    # Create the database.
    db.create_all()


    # Create schema
    class PersonSchema(Schema):
        class Meta:
            type_ = 'person'
            self_view = 'person_detail'
            self_view_kwargs = {'id': '<id>'}
            self_view_many = 'person_list'

        id = fields.Integer(as_string=True, dump_only=True)
        name = fields.String()

        @pre_load
        def remove_id_before_deserializing(self, data, **kwargs):
            """
            We don't want to allow editing ID on POST / PATCH

            Related issues:
            https://github.com/AdCombo/flask-combo-jsonapi/issues/34
            https://github.com/miLibris/flask-rest-jsonapi/issues/193
            """
            if 'id' in data:
                del data['id']
            return data

    # Create resource managers
    class PersonList(ResourceList):
        schema = PersonSchema
        data_layer = {
            'session': db.session,
            'model': Person,
        }


    class PersonDetail(ResourceDetail):
        schema = PersonSchema
        data_layer = {
            'session': db.session,
            'model': Person,
        }


    # Create the API object
    api = Api(app)
    api.route(PersonList, 'person_list', '/persons')
    api.route(PersonDetail, 'person_detail', '/persons/<int:id>')

    # Start the flask loop
    if __name__ == '__main__':
        app.run()


This example provides the following API structure:

========================  ======  =============  ===========================
URL                       method  endpoint       Usage
========================  ======  =============  ===========================
/persons                  GET     person_list    Get a collection of persons
/persons                  POST    person_list    Create a person
/persons/<int:person_id>  GET     person_detail  Get person details
/persons/<int:person_id>  PATCH   person_detail  Update a person
/persons/<int:person_id>  DELETE  person_detail  Delete a person
========================  ======  =============  ===========================


`More detailed example in the docs <https://flask-combo-jsonapi.readthedocs.io/en/stable/minimal_api_example.html>`_


Flask-COMBO-JSONAPI vs `Flask-RESTful`_
==========================================================================================

* In contrast to Flask-RESTful, Flask-COMBO-JSONAPI provides a default implementation of get, post, patch and delete methods around a strong specification JSONAPI 1.0. Thanks to this you can build REST API very quickly.
* Flask-COMBO-JSONAPI is as flexible as Flask-RESTful. You can rewrite every default method implementation to make custom work like distributing object creation.

Flask-COMBO-JSONAPI vs `Flask-Restless`_
===========================================================================================

* Flask-COMBO-JSONAPI is a real implementation of JSONAPI 1.0 specification. So in contrast to Flask-Restless, Flask-COMBO-JSONAPI forces you to create a real logical abstration over your data models with `Marshmallow <https://marshmallow.readthedocs.io/en/latest/>`_. So you can create complex resource over your data.
* In contrast to Flask-Restless, Flask-COMBO-JSONAPI can use any ORM or data storage through the data layer concept, not only `SQLAlchemy <http://www.sqlalchemy.org/>`_. A data layer is a CRUD interface between your resource and one or more data storage so you can fetch data from any data storage of your choice or create resource that use multiple data storages.
* Like I said previously, Flask-COMBO-JSONAPI is a real implementation of JSONAPI 1.0 specification. So in contrast to Flask-Restless you can manage relationships via REST. You can create dedicated URL to create a CRUD API to manage relationships.
* Plus Flask-COMBO-JSONAPI helps you to design your application with strong separation between resource definition (schemas), resource management (resource class) and route definition to get a great organization of your source code.
* In contrast to Flask-Restless, Flask-COMBO-JSONAPI is highly customizable. For example you can entirely customize your URLs, define multiple URLs for the same resource manager, control serialization parameters of each method and lots of very useful parameters.
* Finally in contrast to Flask-Restless, Flask-COMBO-JSONAPI provides a great error handling system according to JSONAPI 1.0. Plus the exception handling system really helps the API developer to quickly find missing resources requirements.

Documentation
=============

Documentation available here: https://flask-combo-jsonapi.readthedocs.io/

Thanks
======

Flask, marshmallow, marshmallow_jsonapi, sqlalchemy, Flask-RESTful and Flask-Restless are awesome projects. These libraries gave me inspiration to create Flask-COMBO-JSONAPI, so huge thanks to authors and contributors.

.. _`Flask-Restless`: https://flask-restless.readthedocs.io/
.. _`Flask-RESTful`: https://flask-restful.readthedocs.io/


docs/changelog.rst
Changelog
*********


**1.1.0**
=========

Enhancements and bug fixes
==========================

* Raise 404 ObjectNotFound when a single resource does not exist `#51`_ #- `@samuelfirst`_
* Improve filter shortcuts `#53`_ #- `@igieon`_
* Ability to use custom query string manager `#54`_ #- `@igieon`_
* Pagination improvement use dict of ints and expect it everywhere `#55`_ #- `@igieon`_
* Enable documented global decorators `#56`_ #- `@igieon`_
* Patch request validates id match before loading schema `#57`_ #- `@igieon`_
* Fix patch nested field partial update `#58`_ #- `@igieon`_


.. _`@samuelfirst`: https://github.com/samuelfirst
.. _`@igieon`: https://github.com/igieon

.. _`#51`: https://github.com/AdCombo/flask-combo-jsonapi/pull/51
.. _`#53`: https://github.com/AdCombo/flask-combo-jsonapi/pull/53
.. _`#54`: https://github.com/AdCombo/flask-combo-jsonapi/pull/54
.. _`#55`: https://github.com/AdCombo/flask-combo-jsonapi/pull/55
.. _`#56`: https://github.com/AdCombo/flask-combo-jsonapi/pull/56
.. _`#57`: https://github.com/AdCombo/flask-combo-jsonapi/pull/57
.. _`#58`: https://github.com/AdCombo/flask-combo-jsonapi/pull/58


examples/api_nested.py
from flask import Flask
from sqlalchemy import UniqueConstraint

from flask_combo_jsonapi import Api, ResourceDetail, ResourceList, ResourceRelationship
from flask_combo_jsonapi.data_layers.alchemy import SqlalchemyDataLayer
from flask_combo_jsonapi.exceptions import ObjectNotFound
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy.orm.exc import NoResultFound
from marshmallow_jsonapi.flask import Schema, Relationship
from marshmallow import Schema as MarshmallowSchema, pre_load
from marshmallow_jsonapi import fields

# Create the Flask application
app = Flask(__name__)
app.config['DEBUG'] = True

# Initialize SQLAlchemy
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:////tmp/api_nested.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
db = SQLAlchemy(app)


# Create data storage
class Person(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String)
    email = db.Column(db.String)
    password = db.Column(db.String)
    tags = db.relationship("PersonTag", cascade="save-update, merge, delete, delete-orphan")
    single_tag = db.relationship(
        "PersonSingleTag",
        uselist=False,
        cascade="save-update, merge, delete, delete-orphan",
    )
    json_data = db.Column(db.JSON)


class Computer(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    serial = db.Column(db.String)
    person_id = db.Column(db.Integer, db.ForeignKey('person.id'))
    person = db.relationship('Person', backref=db.backref('computers'))


class PersonTag(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    person_id = db.Column(db.Integer, db.ForeignKey('person.id'), index=True)
    key = db.Column(db.String)
    value = db.Column(db.String)

    __table_args__ = (
        UniqueConstraint(
            'person_id',
            'key',
            'value',
            name='_person_key_value'
        ),
    )


class PersonSingleTag(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    person_id = db.Column(db.Integer, db.ForeignKey('person.id'), index=True)
    key = db.Column(db.String)
    value = db.Column(db.String)

    __table_args__ = (
        UniqueConstraint(
            'person_id',
            'key',
            'value',
            name='_person_key_value'
        ),
    )


db.create_all()


# Create schema
class PersonTagSchema(MarshmallowSchema):
    class Meta:
        type_ = 'person_tag'

    id = fields.String(dump_only=True, load_only=True)
    key = fields.String()
    value = fields.String()


class PersonSingleTagSchema(MarshmallowSchema):
    class Meta:
        type_ = 'person_single_tag'

    id = fields.String(dump_only=True, load_only=True)
    key = fields.String()
    value = fields.String()


# Create logical data abstraction (same as data storage for this first example)
class PersonSchema(Schema):
    class Meta:
        type_ = 'person'
        self_view = 'person_detail'
        self_view_kwargs = {'id': '<id>'}
        self_view_many = 'person_list'

    id = fields.Integer(as_string=True, dump_only=True)
    name = fields.String(required=True)
    email = fields.Email(load_only=True)
    display_name = fields.Function(lambda obj: "{} <{}>".format(obj.name.upper(), obj.email))
    tags = fields.Nested(PersonTagSchema, many=True)
    single_tag = fields.Nested(PersonSingleTagSchema, many=False)
    json_data = fields.Dict(dump_only=True)
    computers = Relationship(
        self_view='person_computers',
        self_view_kwargs={'id': '<id>'},
        related_view='computer_list',
        related_view_kwargs={'id': '<id>'},
        many=True,
        schema='ComputerSchema',
        type_='computer',
    )

    @pre_load
    def remove_id_before_deserializing(self, data, **kwargs):
        """
        We don't want to allow editing ID on POST / PATCH

        Related issues:
        https://github.com/AdCombo/flask-combo-jsonapi/issues/34
        https://github.com/miLibris/flask-rest-jsonapi/issues/193
        """
        if 'id' in data:
            del data['id']
        return data


class ComputerSchema(Schema):
    class Meta:
        type_ = 'computer'
        self_view = 'computer_detail'
        self_view_kwargs = {'id': '<id>'}

    id = fields.Integer(as_string=True, dump_only=True)
    serial = fields.String(required=True)
    owner = Relationship(
        attribute='person',
        self_view='computer_person',
        self_view_kwargs={'id': '<id>'},
        related_view='person_detail',
        related_view_kwargs={'computer_id': '<id>'},
        schema='PersonSchema',
        type_='person',
    )

    @pre_load
    def remove_id_before_deserializing(self, data, **kwargs):
        """
        We don't want to allow editing ID on POST / PATCH

        Related issues:
        https://github.com/AdCombo/flask-combo-jsonapi/issues/34
        https://github.com/miLibris/flask-rest-jsonapi/issues/193
        """
        if 'id' in data:
            del data['id']
        return data


# Create resource managers
class PersonList(ResourceList):
    schema = PersonSchema
    data_layer = {
        'session': db.session,
        'model': Person,
    }


class PersonDetailSqlalchemyDataLayer(SqlalchemyDataLayer):

    def before_get_object(self, view_kwargs):
        if not view_kwargs.get('computer_id'):
            return
        try:
            computer = self.session.query(Computer).filter_by(
                id=view_kwargs['computer_id'],
            ).one()
        except NoResultFound:
            raise ObjectNotFound(
                "Computer: {} not found".format(view_kwargs['computer_id']),
                source={'parameter': 'computer_id'},
             )
        else:
            if computer.person is not None:
                view_kwargs['id'] = computer.person.id
            else:
                view_kwargs['id'] = None


class PersonDetail(ResourceDetail):
    schema = PersonSchema
    data_layer = {
        'session': db.session,
        'model': Person,
        'class': PersonDetailSqlalchemyDataLayer,
    }


class PersonRelationship(ResourceRelationship):
    schema = PersonSchema
    data_layer = {
        'session': db.session,
        'model': Person
    }


class RelatedComputersSqlalchemyDataLayer(SqlalchemyDataLayer):

    def query(self, view_kwargs):
        query_ = self.session.query(Computer)
        if view_kwargs.get('id') is not None:
            try:
                self.session.query(Person).filter_by(id=view_kwargs['id']).one()
            except NoResultFound:
                raise ObjectNotFound(
                    "Person: {} not found".format(view_kwargs['id']),
                    source={'parameter': 'id'},
                )
            else:
                query_ = query_.join(Person).filter(Person.id == view_kwargs['id'])
        return query_

    def before_create_object(self, data, view_kwargs):
        if view_kwargs.get('id') is not None:
            person = self.session.query(Person).filter_by(id=view_kwargs['id']).one()
            data['person_id'] = person.id


class ComputerList(ResourceList):
    schema = ComputerSchema
    data_layer = {
        'session': db.session,
        'model': Computer,
        'class': RelatedComputersSqlalchemyDataLayer,
    }


class ComputerDetail(ResourceDetail):
    schema = ComputerSchema
    data_layer = {
        'session': db.session,
        'model': Computer,
    }


class ComputerRelationship(ResourceRelationship):
    schema = ComputerSchema
    data_layer = {
        'session': db.session,
        'model': Computer,
    }


# Create endpoints
api = Api(app)

api.route(PersonList, 'person_list', '/persons')
api.route(PersonDetail, 'person_detail', '/persons/<int:id>', '/computers/<int:computer_id>/owner')
api.route(PersonRelationship, 'person_computers', '/persons/<int:id>/relationships/computers')
api.route(ComputerList, 'computer_list', '/computers', '/persons/<int:id>/computers')
api.route(ComputerDetail, 'computer_detail', '/computers/<int:id>')
api.route(ComputerRelationship, 'computer_person', '/computers/<int:id>/relationships/owner')

if __name__ == '__main__':
    # Start application
    app.run(debug=True)
